### PR TITLE
Decoupled Weight Decay Regularization in optimizers (added adamw and sgdw among others)

### PR DESCRIPTION
**Pull request by kashif**
I have added the adamw and sgdw flags to the appropriate optimizers rather than their own ones for issue #3790

~~Instead of defining new optimizers as in  PR #3740 I am fixing the weight decay in the appropriate optimizers.~~

~~My only issue is that now the comparison tests between the older legacy optimizer and this one fails.~~

### CODE REVIEW

**Comment by HaraldKorneliussen** [comment]
It looks like it's just the flake test complaining about extra whitespace.
And the way to fix the failing "equivalent to lua torch"-test, I guess is to fix it in lua torch first? :)

**Comment by kashif** [comment]
@HaraldKorneliussen thanks! Yes that is the question, if I should also change the legacy code or remove the test?

File: torch/optim/adadelta.py
Comment (vishwakftw) [inline torch/optim/adadelta.py]: Can't this be:
```python
if group['weight_decay'] != 0:
   p.data.add_(-group['weight_decay'], p.data)
   p.data.add_(-group['lr'], delta)
```

**Review by kashif — COMMENTED** [review]
ok I removed the unnecessary clones

File: torch/optim/adadelta.py
Comment (vishwakftw) [reply torch/optim/adadelta.py]: `add_(scale, tensor)` is an inplace operation over the tensor that calls this routine.

**Comment by lucasb-eyer** [comment]
I would like to see this merged, as I think that paper's result was an important one. I also prefer "fixing" the existing optimizers as in this PR over the huge copy-paste in the other PR.

As it currently is implemented, I have never encountered a case where weight-decay has helped me in any of the non-SGD/Momentum optimizers, nor has anyone I have talked to.

That being said, we might want to keep the original behaviour by default, at the very least for SGD/Momentum. Imagine all the training codes where the performance changes (for the better or the worse) after an update, and how near-impossible it would be to find out that this PR is why! Then, read the paper, and fix the factors accordingly. So I'd rather see a new flag for it, something like `separate_weight_decay=False` by default.

Or maybe print a (one-time per session) warning whenever the optimizer is created with non-zero `weight_decay` for the next handful of versions.

**Comment by apaszke** [comment]
I'm a bit concerned about this PR, because it completely changes how optimizers work, and may hurt reproducibility. Any opinions @soumith?

**Comment by kashif** [comment]
I agree @apaszke it will cause issues with reproducibility. What is the pytorch design principle? to stay backwards compatible? I could add a flag that switches from default to this behaviour as @lucasb-eyer suggests...

**Comment by lucasb-eyer** [comment]
@apaszke you are right in that it would be different from most (all?) popular frameworks. I haven't seen enough confirmation of its effectiveness in the wild (have you, @kashif?) so maybe the conservative `separate_weight_decay=False` is the better thing to do here. But no matter how, I'd really like to have this in PyTorch.

Or be the bold Apple of DL and move to Doing It Right(TM), just with a big fat warning [as I mentioned here](https://github.com/pytorch/pytorch/pull/4429#issuecomment-357508516) :smile:

**Comment by kashif** [comment]
@jpeg729 the pytorch api makes it seem like I am not multiplying the learning rate, especially the inplace APIs, but if see the api call bellow that, then you will see that the learning rate gets multiplied etc. as in the paper... well not quite, your kind comment made me realise I have a bug... I have a negative sign too many! Fixing that now! Thanks and hope it helps!

**Comment by jpeg729** [comment]
@kashif I beg to differ.
For example, from your modified code for SGD. I have added comments according to my understanding of the pytorch inplace apis.

                if weight_decay != 0:
                    p.data.add_(-weight_decay, p.data) # p.data = p.data + weight_decay * p.data

                p.data.add_(-group['lr'], d_p) # p.data = p.data + lr * d_p

Only d_p gets multiplied by the learning rate, not the weight_decay update.
As far as I can see, none of the optimizers multiply the weight_decay by the learning rate.

File: torch/optim/adagrad.py
Comment (jpeg729) [inline torch/optim/adagrad.py]: The new weight decay update is added directly to p.data, so I think sparse gradients are no longer a problem for weight_decay.

**Comment by kashif** [comment]
@jpeg729 ok so I fixed the negatives:

```
if weight_decay != 0:
    p.data.add_(-weight_decay, p.data) # p.data = p.data - weight_decay * p.data

p.data.add_(-group['lr'], d_p) # p.data = p.data - lr * d_p = p.data -lr * d_p - weight_decay * p.data
```

which is essentially line 9 in Algorithm 1 of the paper. Could it be that you are thinking of the `eta` in the paper as the learning rate? `eta` in the paper is the learning rate schedular which here is just 1.

**Comment by jpeg729** [comment]
@kashif You are right. `eta` is the Scheduler.
But, I think that there is a problem here that requires some thought.
In the paper at time t the effective learning rate is `eta * alpha`, and the effective weight_decay is `eta * w`.

So if someone uses a learning rate scheduler with these optimisers, _then they have to manually apply a similar scheduler to the weight decay_ in order to replicate the methods in the paper. Otherwise the grad updates would decay but not the weight decay updates, and eventually the grad updates would be swamped by the weight decay updates.

One fix would be to add `param_group['eta']` to each optimiser and change the schedulers to modify `param_group['eta']` instead of `param_group['lr']`.

**Comment by kashif** [comment]
@jpeg729 yes you are right, at the moment the schedular will only change the learning rate and not the weight decay parameter... let me see how hard it will be to add that.. but as you can read from the comments above I dont think these changes are going to be accepted so not sure how much its worth investing time on this... I might close this PR soon.

**Comment by lucasb-eyer** [comment]
@kashif if this PR does not get merged (likely), do you plan to create a separate repo that trackes optim but with these changes? Something like `optimw`, or will you just let it die and not use these changes anymore yourself?

**Comment by kashif** [comment]
@lucasb-eyer good idea... I will move it over and leave a comment here to the new location as soon as I get some free time...

**Comment by kashif** [comment]
@ezyang the tests will fail since the comparison with the old optimizers will not be the same...

**Comment by snowzjy** [comment]
Can adamw be used  now？

**Comment by kashif** [comment]
@silver-rush no i don't think that will work... Let me double check it though...

**Comment by kashif** [comment]
@apaszke I have now added a flag that by default does the original L2 weight regularisation so by default nothing is different.

**Comment by kashif** [comment]
@jpeg729 and @silver-rush I have updated the LR Schedulers to also schedule the weight decay factor when using the fixed weight decay in the optimizer.

**Comment by kashif** [comment]
@weiyangfb can you kindly have another look? Thanks!

**Comment by kashif** [comment]
@apaszke @soumith can you kindly have a look? Would love to see this merged before the 1.0 release thanks!

File: torch/optim/adadelta.py
Comment (MicaelCarvalho) [inline torch/optim/adadelta.py]:
```suggestion
                    p.data.add_(-group['weight_decay'] * group['lr'], p.data)
```

In L2, the weight decay is stored inside the grad variable, which is later multiplied by the learning rate. This new formulation is not multiplying it by the learning rate (see Algorithms 1 and 2 in the paper).

File: torch/optim/adagrad.py
Comment (MicaelCarvalho) [inline torch/optim/adagrad.py]:
```suggestion
                    p.data.add_(-group['weight_decay'] * group['lr'], p.data)
```

In L2, the weight decay is stored inside the grad variable, which is later multiplied by the learning rate. This new formulation is not multiplying it by the learning rate (see Algorithms 1 and 2 in the paper).

File: torch/optim/adam.py
Comment (MicaelCarvalho) [inline torch/optim/adam.py]:
```suggestion
                    p.data.add_(-group['weight_decay'] * group['lr'], p.data)
```

In L2, the weight decay is stored inside the grad variable, which is later multiplied by the learning rate. This new formulation is not multiplying it by the learning rate (see Algorithms 1 and 2 in the paper).

File: torch/optim/adamax.py
Comment (MicaelCarvalho) [inline torch/optim/adamax.py]:
```suggestion
                    p.data.add_(-group['weight_decay'] * group['lr'] , p.data)
```

In L2, the weight decay is stored inside the grad variable, which is later multiplied by the learning rate. This new formulation is not multiplying it by the learning rate (see Algorithms 1 and 2 in the paper).

File: torch/optim/asgd.py
Comment (MicaelCarvalho) [inline torch/optim/asgd.py]:
```suggestion
                    p.data.add_(-group['weight_decay'] * group['lr'], xold)
```

In L2, the weight decay is stored inside the grad variable, which is later multiplied by the learning rate. This new formulation is not multiplying it by the learning rate (see Algorithms 1 and 2 in the paper).

File: torch/optim/rmsprop.py
Comment (MicaelCarvalho) [inline torch/optim/rmsprop.py]:
```suggestion
                        p.data.add_(-group['weight_decay'] * group['lr'], p.data)
```

In L2, the weight decay is stored inside the grad variable, which is later multiplied by the learning rate. This new formulation is not multiplying it by the learning rate (see Algorithms 1 and 2 in the paper).

File: torch/optim/sgd.py
Comment (MicaelCarvalho) [inline torch/optim/sgd.py]:
```suggestion
                    p.data.add_(-weight_decay * group['lr'], p.data)
```

In L2, the weight decay is stored inside the grad variable, which is later multiplied by the learning rate. This new formulation is not multiplying it by the learning rate (see Algorithms 1 and 2 in the paper).

**Review by MicaelCarvalho — CHANGES_REQUESTED** [review]
Edit: Invalid suggestion, removed. Thanks for the clarification @kashif

File: torch/optim/adam.py
Comment (kashif) [reply torch/optim/adam.py]: I will reply to this comment only since the rest are the same... yes in the l2 reg you are right the weight decay is in the grad and it gets multiplied by the learning rate... and this is the default behaviour when `l2_reg = True`. But  in this block we are doing what the paper suggests which is the green bit in Algorithm 2 of the paper which only happens when `l2_reg=False`. Hope that clears the confusion?

Comment (MicaelCarvalho) [reply torch/optim/adam.py]: Algorithm 2 has this term inside parenthesis, it _is_ being multiplied by the learning rate over there, not here... ;)

Comment (kashif) [reply torch/optim/adam.py]: ah you mean the `\eta_t` in Algorithm 2? Remember `\eta_t` is the learning rate schedular not the learning rate, the learning rate in the paper is `\alpha`... hope that helps

Comment (MicaelCarvalho) [reply torch/optim/adam.py]: I see, thank you for the clarification. :-)

I find it weird to decouple weight decay from learning rate. In my understanding, their claims would remain valid by applying the learning rate to the new weight decay implementation as well, without having weight decay going rogue.

This can lead to so many confusing results, because if one changes the learning rate, the decay remains just as strong as it was before. People would need to beware of both and adjust both accordingly for different experiments.

Comment (kashif) [reply torch/optim/adam.py]: ok then closing this as resolved...

Comment (kashif) [reply torch/optim/adam.py]: yes so the schedulers in this pull request also schedule the weight decay by the same `\eta_t` as the learning rate... so hopeful you don't need to think about adjusting the weight decay yourself.

File: torch/optim/adam.py
Comment (ogrisel) [inline torch/optim/adam.py]: The [ICLR 2019 update of the AdamW paper](https://arxiv.org/abs/1711.05101) was renamed to "Decoupled Weight Decay Regularization" (which I find a better name).

File: torch/optim/adam.py
Comment (ogrisel) [reply torch/optim/adam.py]: Maybe it would be worth to even issue a `FutureWarning` stating that `l2_reg` will be set to `False` by default in a future version?

```python
def __init__(self, params, lr=1e-3, betas=(0.9, 0.999), eps=1e-8,
             weight_decay=0, l2_reg="warn", amsgrad=False):
    if l2_reg == "warn":
        warnings.warn("l2_reg will be set to False by default in a future Pytorch version."
                      " Set l2_reg=False explicit to silence this warning.", FutureWarning)
        l2_reg = True  # backward compatibility
    ...
```

File: torch/optim/adam.py
Comment (ogrisel) [inline torch/optim/adam.py]: The appendix of the new version of [the paper](https://arxiv.org/abs/1711.05101) recommends using a "normalized" parametrization of the weight decay that seems to be more stable across datasets and batch sizes.

The problem is that this normalization needs to be aware of the mini batch size, the full batch sizes and the number of epochs when using the cosine annealing of the learning rate.

Maybe this is too much coupling with external information though.

File: torch/optim/adamax.py
Comment (ogrisel) [inline torch/optim/adamax.py]: Here again this needs to be updated.

File: torch/optim/asgd.py
Comment (ogrisel) [inline torch/optim/asgd.py]: Idem.

File: torch/optim/sgd.py
Comment (ogrisel) [inline torch/optim/sgd.py]: Idem.

**Review by ogrisel — COMMENTED** [review]
Here are some comments:

File: torch/optim/adam.py
Comment (kashif) [reply torch/optim/adam.py]: Thanks @ogrisel for your kind suggestions, I appreciate you taking time to do this. I have fixed the name of the paper everywhere. 

Regarding the recommendation of the default value of the `l2_reg` flag, I am happy to change it but it depends on the PyTorch maintainers... I suspect they are busy with other stuff, but hopefully they reply, this issue has been open now for a year so I am not too hopeful 😿 

If you have time can you also review the changes in the schedular, since now if we use this method, the learning rate and weight decay variables are decoupled and I need to schedule them both and that is a bit messy... Any suggestions to simplify that aspect will be great! Thanks!

Comment (ogrisel) [reply torch/optim/adam.py]: With respect to getting reviewers attention, I believe it would be worth taking the time to try to reproduce some of the experiments of the AdamW paper in gist and then publish a screenshot of the train loss and validation accuracy convergence plots in a comments in this PR (or even include those in the description of the PR itself).

File: torch/optim/adam.py
Comment (acburigo) [inline torch/optim/adam.py]: Not obviously clear by the text which option is which. Maybe state it, like
```
            l2_reg (boolean, optional): whether to use the original L2
            weight regularization (l2_reg=True) or the weight decay method from the paper
            `Decoupled Weight Decay Regularization` (l2_reg=False) (default: True)
```

File: torch/optim/adam.py
Comment (loshchil) [reply torch/optim/adam.py]: @ogrisel I agree that it would be better to go for the simpler version, i.e., without the normalization based on the mini batch size and full batch size as described in the paper.

**Comment by loshchil** [comment]
> @jpeg729 As a workaround using the current code here, will it correctly apply scheduling on the weight decay if I set
> 
> `weight_decay = actual_weight_decay / initial_lr`
> and correspondingly, in the optimizer,
> `p.data.add_(-group['weight_decay'] * group['lr'], p.data)`
> 
> Such that the initial learning rate get cancelled out and is left with `eta`?

I didn't try it but I think that you are right @hkchengrex  , this should work and implement eta as described in the paper. Otherwise, the current code has the issue described above by @jpeg729 
At first, it might seem unusual that weight decay is scheduled. However, a closer look at the original SGD with L2-based weight decay will reveal that weight decay is scheduled there as well because the weight decay part is first combined with the loss-based gradient and then multiplied by learning rate which is  usually scheduled. 
Please note note that the optimal value of weight decay in SGDW is likely to be different by a factor of lr from the one of SGD with L2, see Section 2 of https://arxiv.org/pdf/1711.05101.pdf.

**Comment by soumith** [comment]
cc: @vincentqb can you review this and get it to completion if it makes sense. Use the guidelines that I shared with you separately.

@kashif i want to apologize for never getting a review to completion, but we now have Vincent with a lot of bandwidth and who is a math and optimization expert. He will help with the review.

**Comment by kashif** [comment]
thanks @soumith I understand the delay. I will try to work with @vincentqb to get it ready for another review!

**Comment by kashif** [comment]
@soumith @vincentqb i see you have merged versions of this optimizer without reviewing my PR pending for more than a year, as you requested 17 days ago in the comment above. I find that disrespectful even if i understood the limited bandwidth you have but to then go ahead and accept copies of this work is  unfair.

**Comment by loshchil** [comment]
@soumith @vincentqb 
It seems that [the new code](https://github.com/pytorch/pytorch/commit/a4b2f3e21316461598fbebfdeb97259d2bbf29fa#diff-5df3e120a54ab4051f77aad451cba1a1R69) does not take into account the implementation issue discussed in the current PR by @kashif (e.g., see my comment above and also [here](https://github.com/pytorch/pytorch/issues/22343#issuecomment-508950827) ). In might be necessary to make users aware that implemented algorithm only partially follows the paper (it follows the main contribution of the paper  that the weight decay update is taken out from the adaptive gradient update but the 'lr' and 'weight_decay' hyperparameters are still coupled in contrast to the paper where a scheduler is used). 

I would like to thank you all and especially @kashif for investing your time to disseminate that work.

**Comment by soumith** [comment]
@kashif it was disrespectful, I fully agree -- sorry about that. It's also not a situation we wanted to be in -- where multiple PRs were pending for the same algorithm, and we had no maintainers for `optim` for like forever. If I could undo it one way or the other I would. I'm banking on @vincentqb being a maintainer who will over time do better (he is still ramping up, and I'm sure he doesn't have full context yet).

@vincentqb can you look at what @loshchil mentioned.

**Comment by kashif** [comment]
@soumith thank you for your kind message and i accept your apology and apologise for my outburst. I too will try to be more understanding. Sleeping over it I realised there was something else bothering me that I needed to work through. Anyways, let me know if I can help in anyway.

**Comment by vincentqb** [comment]
@kashif Thank you for your understanding. I appreciate your contributions -- and everyone else's in this pull request -- and I also want to apologize. I was in the process of going through this pull request.

The two key contributions in AdamW and SGDW seems to be:
1. Weight decay done through the parameter update directly theta (instead of through the parameter gradient g: line 6 vs 9 in algorithm 1, line 6 vs 12 in algorithm 2),
2. The definition of a new ScheduleMultiplier `eta(t)` replacing LRScheduler `lr(t)`. In that case, the implied relation between ScheduleMultiplier `eta(t)` and LRScheduler `lr(t)` is `lr(t) = lambda*alpha*eta(t)`, with lambda and alpha two constant.

Based on the discussions above, I am also understanding that we could update lambda and alpha independently instead of eta. Thus effectively having two separate schedulers, one for the momentum update and one for parameter update. This would be why 2 above is referred to as "decoupled weight decay". Is this what you mean @loshchil ?

If that is the case, my initial plan was the following. Given that we do not have a separate momentum scheduler concept yet #22343, the initial plan was to bring in AdamW and SGDW with "coupled" learning rate and weight decay (and thus using the same scheduler). However, based on the discussion above, it appears AdamW and SGDW have this decoupling implied to the user, and not having the decoupling would bring confusion.

Thoughts?

**Pull request closed without merging**
